### PR TITLE
Add a synced use enhanced color checkbox to C# and Basic advanced options.

### PR DESCRIPTION
**Pull request by JoeRobich**
Moves unsaved option changes out of the WPF Bindings into an OptionStore class. This allows for the C# and Basic option checkboxes to stay in sync.

<details><summary>Ask Mode template</summary>

### Customer scenario

The "Use enhanced colors for managed languages" checkbox will be removed from the Preview Features option page with this PR https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/164798. This adds back the "Use enhanced colors for managed languages" checkbox into pages under C# and Basic.

### Bugs this fixes

N/A

### Workarounds, if any

N/A

### Risk

Minimal

### Performance impact

None.

### Is this a regression from a previous update?

N/A

### Root cause analysis

N/A

### How was the bug found?

N/A

</details>

### CODE REVIEW

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 1]
```diff
@@ -1,23 +1,39 @@
...
+        private static bool s_needsLoadOnNextActivate = true;
+
+        protected abstract AbstractOptionPageControl CreateOptionPage(IServiceProvider serviceProvider, OptionStore optionStore);
 
         protected AbstractOptionPageControl pageControl;
-        private bool _needsLoadOnNextActivate = true;
 
         private void EnsureOptionPageCreated()
         {
+            if (s_optionStore == null)
```
Comment (ryzngard) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Are there any concerns with threading/race conditions here?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 2]
```diff
@@ -1,23 +1,39 @@
...
+            {
+                var componentModel = (IComponentModel)this.Site.GetService(typeof(SComponentModel));
+                var workspace = componentModel.GetService<VisualStudioWorkspace>();
+                s_optionService = workspace.Services.GetService<IOptionService>();
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: I'm guessing this should be directly importing `IGlobalOptionService`? The comment does say:

https://github.com/dotnet/roslyn/blob/1e3585793758cafd42ca9ba7bda1eab88da681d6/src/Workspaces/Core/Portable/Options/IGlobalOptionService.cs#L13-L14

But I think you are the exception in this place. Usually the assumption is people shouldn't be reading global options _directly_ because a workspace can provide customized options. But in this case this is _configuring_ the global options themselves.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 3]
```diff
@@ -32,13 +48,18 @@ protected override System.Windows.UIElement Child
...
-            if (_needsLoadOnNextActivate)
+            EnsureOptionPageCreated();
+
+            if (s_needsLoadOnNextActivate)
             {
-                EnsureOptionPageCreated();
-                pageControl.LoadSettings();
+                // Reset the option store to the current state of the options.
+                s_optionStore.SetOptions(s_optionService.GetOptions());
+                s_optionStore.SetRegisteredOptions(s_optionService.GetRegisteredOptions());
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Why was this necessary? The list of registered options really shouldn't be able to change, since that's coming from MEF.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 4]
```diff
@@ -1,23 +1,39 @@
...
     internal abstract class AbstractOptionPage : UIElementDialogPage
     {
-        protected abstract AbstractOptionPageControl CreateOptionPage(IServiceProvider serviceProvider);
+        private static IOptionService s_optionService;
+        private static OptionStore s_optionStore;
+        private static bool s_needsLoadOnNextActivate = true;
+
+        protected abstract AbstractOptionPageControl CreateOptionPage(IServiceProvider serviceProvider, OptionStore optionStore);
 
         protected AbstractOptionPageControl pageControl;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Not related to your change but this makes it even more confusing: can you rename this to `_pageControl` so we don't confuse it with a local?

File: src/VisualStudio/Core/Impl/Options/OptionBinding.cs [hunk 5]
```diff
@@ -1,34 +1,41 @@
...
-        public OptionBinding(IOptionService optionService, Option<T> key)
+        public event PropertyChangedEventHandler PropertyChanged;
+
+        public OptionBinding(OptionStore optionStore, Option<T> key)
         {
-            _optionService = optionService;
+            _optionStore = optionStore;
             _key = key;
+
+            _optionStore.OptionChanged += (sender, e) =>
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/OptionBinding.cs]: Does this ever get unsubscribed? We don't re-create these if I dismiss and reopen the pages, correct?

File: src/VisualStudio/Core/Impl/Options/OptionStore.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    /// <summary>
+    /// This class is intended to be used by Option pages. It will provide access to an options
+    /// from an optionset but will not persist changes automatically.
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/OptionStore.cs]: Make the OptionSet reference a cref.

File: src/VisualStudio/Core/Impl/Options/OptionStore.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+            OnOptionChanged(new OptionKey(option, language));
+        }
+
+        public IEnumerable<IOption> GetRegisteredOptions()
+        {
+            return _registeredOptions;
+        }
+
+        public void SetOptions(OptionSet optionSet)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/OptionStore.cs]: This doesn't re-raise that options have potentially changed? Thinking about a scenario like this:

1. User previously opens Tools > Options, changes some stuff. Closes it.
2. User changes setting on another machine; it syncs and updates the global option service.
3. User opens the dialog again.
4. The `s_optionStore.SetOptions(s_optionService.GetOptions());` line runs in the option page guts.

What's causing the WPF to refresh the binding in this case?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs [hunk 8]
```diff
@@ -177,4 +188,19 @@ public class RadioButtonCheckedConverter : IValueConverter
             return value.Equals(true) ? parameter : Binding.DoNothing;
         }
     }
+
+    public class CheckBoxCheckedToIntConverter : IValueConverter
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: Do I dare ask what this is?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs [hunk 9]
```diff
@@ -72,9 +83,9 @@ protected void BindToOption(CheckBox checkbox, PerLanguageOption<bool> optionKey
         {
             var binding = new Binding()
             {
-                Source = new PerLanguageOptionBinding<bool>(OptionService, optionKey, languageName),
+                Source = new PerLanguageOptionBinding<bool>(OptionStore, optionKey, languageName),
                 Path = new PropertyPath("Value"),
-                UpdateSourceTrigger = UpdateSourceTrigger.Explicit
+                UpdateSourceTrigger = UpdateSourceTrigger.Default
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: Can we just delete all the UpdateSourceTriggers setting since we're now setting it to default?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs [hunk 10]
```diff
@@ -51,21 +72,28 @@ public override void LoadSettingsFromStorage()
...
             // next time
-            _needsLoadOnNextActivate = true;
+
+            s_needsLoadOnNextActivate = pageControl != null;
         }
 
         public override void SaveSettingsToStorage()
         {
             EnsureOptionPageCreated();
-            pageControl.SaveSettings();
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: I'm guessing you can delete SaveSettings now that it's not being used anymore?

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs [hunk 11]
```diff
@@ -177,4 +188,19 @@ public class RadioButtonCheckedConverter : IValueConverter
             return value.Equals(true) ? parameter : Binding.DoNothing;
         }
     }
+
+    public class CheckBoxCheckedToIntConverter : IValueConverter
+    {
+        public object Convert(object value, Type targetType, object parameter,
+            System.Globalization.CultureInfo culture)
+        {
+            return !value.Equals(-1);
```
Comment (ryzngard) [inline src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: nit: It probably doesn't matter at all, but why is this different from Convert.ToInt32? https://docs.microsoft.com/en-us/dotnet/api/system.convert.toint32?redirectedfrom=MSDN&view=netframework-4.7.2#System_Convert_ToInt32_System_Boolean_

File: src/VisualStudio/Core/Impl/Options/OptionBinding.cs [hunk 12]
```diff
@@ -1,34 +1,41 @@
...
+            {
+                if (e.Option == _key)
+                {
+                    PropertyChanged?.Raise(this, new PropertyChangedEventArgs(nameof(Value)));
```
Comment (ryzngard) [inline src/VisualStudio/Core/Impl/Options/OptionBinding.cs]: Are we raising `PropertyChanged` here because option store setting isn't synchronous? Otherwise we could just raise in the setter

File: src/VisualStudio/CSharp/Impl/Options/AdvancedOptionPageControl.xaml [hunk 13]
```diff
@@ -112,7 +112,15 @@
             </StackPanel>
           </GroupBox>
 
-          <GroupBox x:Uid="ExtractMethodGroupBox"
```
[the remark this answers is not in the document]
Comment (JoeRobich) [reply src/VisualStudio/CSharp/Impl/Options/AdvancedOptionPageControl.xaml]: The indentation is inconsistent in this file. I tried not to disturb the existing indentation and let the .editorconfig do its job with the code I added.

File: src/VisualStudio/Core/Impl/Options/OptionBinding.cs (hunk 12 above)
Comment (ryzngard) [reply src/VisualStudio/Core/Impl/Options/OptionBinding.cs]: oh! Nm, the value could change outside of this binding. I get it now :)

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs (hunk 3 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Ah, I will remove it then. I was just trying to ensure the option store was entirely reset.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs (hunk 11 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: I will add a comment. This was to keep compatibility with the setting used in the Preview Features page. That value is an integer where -1 was disabled, 0 was use the flight information, and 1 was enabled.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs (hunk 1 above)
Comment (ryzngard) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPage.cs]: Talked offline. These are UI objects so they implicitly have thread affinity

File: src/VisualStudio/Core/Impl/Options/OptionBinding.cs (hunk 5 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/OptionBinding.cs]: The pages that create the bindings are create once and live for the lifetime of the application.

File: src/VisualStudio/Core/Impl/Options/OptionStore.cs (hunk 7 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/OptionStore.cs]: Currently we are still executing the `AbstractOptionPageControl.LoadSettings` when a page is activated. I agree that we should remove the `LoadSettings` call and instead fire `OptionChanged` when `SetOptions` is called.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs (hunk 8 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: I should have commented. I'll add a comment.

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs (hunk 11 above)
Comment (ryzngard) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: Ah, so this may not even be correct for all options? Yea, maybe a comment or rename is warranted here to make sure the intended use is understood

File: src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs (hunk 8 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Impl/Options/AbstractOptionPageControl.cs]: This was to keep compatibility with the setting used in the Preview Features page. That value is an integer where -1 was disabled, 0 was use the flight information, and 1 was enabled.
